### PR TITLE
fix(#6467): show final assistant report with tool-limit status card

### LINKED ISSUE #6467: Tool-limit status card hides a non-empty final assistant report
> **Disclosure:** This report was investigated, written and submitted by Hermes Agent through the `mac-wifi` GitHub account, with the account owner's explicit authorisation.

## Bug description

When a max-iteration turn has both:

- a non-empty, useful assistant final response; and
- a WebUI `_statusCard` marking `tool_limit_reached`,

the normal transcript renderer displays only the status card. The assistant's final response is present in the persisted message but is not rendered.

This makes a successfully completed or usefully summarised task look like a bare failure. Other clients that render the persisted `content` without applying WebUI's `_statusCard` precedence show the final report correctly.

This appears to be a frontend follow-up/regression after #3821 and #5494 / #5717. Those changes correctly preserve the graceful final response and terminal metadata, but the current normal renderer treats the two as mutually exclusive.

## Reproduction

1. Run a tool-heavy task until `agent.max_turns` is reached.
2. Allow Hermes Agent's graceful finalisation call to return a non-empty assistant summary.
3. Confirm the persisted assistant message has the equivalent of:

```json
{
  "role": "assistant",
  "content": "## Task completed\n\nA useful non-empty final report...",
  "_terminal_state": "tool_limit_reached",
  "_terminal_reason": "max_iterations",
  "_statusCard": {
    "title": "Tool iteration limit reached",
    "subtitle": "Stopped because the tool iteration limit was reached."
  }
}
```

4. Reload that session in the normal WebUI transcript renderer.

## Actual behaviour

Only the **Tool iteration limit reached** status card is displayed. The non-empty assistant report is hidden.

## Expected behaviour

Display both:

1. the terminal-status card, so the user knows the configured iteration cap fired; and
2. the non-empty assistant final report, so the user can see what was accomplished and what remains.

This matches the acceptance criterion in #3821: "A run with a usable assistant summary displays that summary as Final Answer plus a clear `tool_limit_reached` status."

## Suspected cause

Current upstream `static/ui.js` contains this normal-renderer branch:

```javascript
const hasVisibleBody=!!(String(content||'').trim()||filesHtml||statusHtml||recoveryHtml);
if(statusHtml){
  seg.insertAdjacentHTML('beforeend', statusHtml);
}else if(hasVisibleBody){
  seg.insertAdjacentHTML('beforeend', `${filesHtml}<div class="msg-body">${bodyHtml}</div>${footHtml}`);
}else if(!(thinkingText&&window._showThinking!==false&&!isSimplifiedToolCalling())){
  seg.classList.add('assistant-segment-anchor');
}
```

Because `statusHtml` is truthy, the `else if (hasVisibleBody)` branch never renders the message body—even when `content` is non-empty.

The transparent/ordered-part path nearby appears to append `statusHtml` and the message body independently, so the defect may be limited to the normal rendering branch.

## Suggested direction

Render the status card and non-empty body independently rather than making them mutually exclusive. Preserve the anchor-only behaviour when there is genuinely no visible body or status.

Please add a frontend regression test for a persisted assistant row containing both non-empty `content` and `_statusCard`.

## Verified against

- Repository: `nesquena/hermes-webui`
- Upstream `master`: `399cd7ab19ce43f77d3c3a68bb575e9eed9cd6af`
- The `if(statusHtml) ... else if(hasVisibleBody)` pattern is present at that revision.
- The underlying Hermes Agent result had `turn_exit_reason=max_iterations_reached`, while its stored assistant final response was non-empty.

No private session content, credentials, domain names or user data are included in this report.

### PR DESCRIPTION
**Pull request by rodboev**
## Thinking Path

- `renderMessages()` already renders card plus body in the ordered-transparent path, so the remaining gap is the ordinary assistant branch only.
- Current `origin/master` still inserts `_statusCard` without the normal body when both are present.
- The fix keeps the existing special branches intact and restores the normal prose-first assistant turn.

## What Changed

- `static/ui.js`: ordinary assistant messages now render the status card before the final answer body instead of replacing it.
- `tests/test_status_command_card.py`: added focused regression coverage for normal-with-card, status-only, and unchanged boundary branches.

## Why It Matters

When a tool-limit turn still produces a useful final answer, the WebUI now shows that answer instead of hiding it behind the status card. The card stays visible, and the assistant prose stays readable in the same turn.

## Verification

The focused status-card regression now passes, including the ordinary-with-card, status-only, body-only, and file-only boundaries. The same seeded ordinary-with-card case drops the final report on `origin/master`. Browser smoke stayed clean on `/`, `/#settings`, and `/#sessions`, and the rendered-ui proof plus screenshots were captured before PR open.

## Risks / Follow-ups

Low risk. The change stays inside the ordinary assistant render branch and preserves the existing special-case paths that already render status cards correctly.

## Upstream

Closes #6467.

## Screenshots

Before, the tool-limit status card replaced the assistant's final report, so the turn stopped at the card.

Before

After, the same turn keeps the status card first and restores the final assistant report directly underneath it.

After

## Model Used

GPT-5.5 via Codex CLI

### CODE REVIEW

File: tests/test_status_command_card.py [hunk 1]
```diff
@@ -81,6 +272,43 @@ def test_render_messages_treats_status_card_as_visible_assistant_content():
...
+    probe = _status_render_probe()
+    ordered = probe["orderedSpecial"]
+    assert ordered["statusCount"] == 1
+    assert ordered["bodyCount"] == 1
+    assert ordered["html"] == "<status-card>limit</status-card><div class=\"msg-body\"><p>Final report</p></div><footer>meta</footer>"
+
+
 def test_status_card_styles_exist():
     assert ".status-card" in STYLE_CSS
     assert ".status-card-grid" in STYLE_CSS
```
[the remark this answers is not in the document]
Comment (rodboev) [reply tests/test_status_command_card.py:314]: Leaving this as-is. This helper is test-local, the focused file still runs quickly, and keeping each probe call self-contained avoids adding shared cached state to a narrow regression harness.

File: tests/test_status_command_card.py [hunk 2]
```diff
@@ -31,6 +38,190 @@ def _function_body(src: str, name: str) -> str:
...
+}}
+
+const ordinaryBlock = extractBlock(
+  "const hasVisibleBody=!!(String(content||'').trim()||filesHtml||recoveryHtml);",
+  "_assistantTurnBlocks(currentAssistantTurn).appendChild(seg);"
+);
+const orderedBlock = extractBlock(
+  "if(isLastTextPart&&statusHtml){{",
+  "blocks.appendChild(orderedSeg);"
+);
```
[the remark this answers is not in the document]
Comment (rodboev) [reply tests/test_status_command_card.py:97]: Agreed that the source markers are brittle, and that's deliberate here. This probe is guarding the exact render seam introduced by the PR, so I want it to fail as soon as that block shape changes instead of silently drifting behind a broader extractor.

**Comment by nesquena-hermes** [comment]
## Summary

Reading the complete `static/ui.js` render path at head and on `origin/master`, plus the full status-card regression file, this is the correct fix for the ordinary assistant branch. The previous predicate counted the status card itself as a visible body, then the `statusHtml` branch emitted only that card. This revision separates card presence from actual body presence and emits both in a deterministic order. I do not see a correctness blocker.

## Code reference

The repaired branch at `static/ui.js:16408-16417` now has the right two independent decisions:

```js
const hasVisibleBody=!!(String(content||'').trim()||filesHtml||recoveryHtml);
if(statusHtml){
  seg.insertAdjacentHTML('beforeend', statusHtml);
  if(hasVisibleBody) seg.insertAdjacentHTML('beforeend', `${filesHtml}<div class="msg-body">${bodyHtml}</div>${footHtml}`);
}
```

Removing `statusHtml` from `hasVisibleBody` is important for the status-only boundary: it prevents an empty `msg-body` from being added just because a card exists. The nested insertion then restores the body only when prose, files, or recovery content exists. The unchanged `else if(hasVisibleBody)` path at `static/ui.js:16416-16417` preserves body-only messages.

The test harness at `tests/test_status_command_card.py:18-63` extracts and evaluates the production branch rather than duplicating a separate implementation. Its cases at `tests/test_status_command_card.py:76-143` cover the regression and the relevant negative space: card plus report, status only, body only, files plus status, and the already-correct ordered-transparent path.

## Diagnosis / recommendation

The change is minimal and is placed at the actual ownership boundary. It does not alter status-card construction, markdown rendering, recovery extraction, or the special transparent-stream branches. The output order is card, files/body, then footer, which matches the intended screenshot and keeps the final report in the same assistant turn.

## Verification

I performed a read-only full-file and base/head diff review; no contributor code was executed from the worktree. The current GitHub rollup is green across browser smoke, both lifecycle jobs, lint, docs, and all Python test shards. Verdict: ready for maintainer merge review.

**Comment by nesquena-hermes** [comment]
Shipped in **exp-v0.52.165**. Thanks @rodboev — and approved by Nathan on the desktop+mobile A/B screenshots.

Gate on current master:
- Full single-process suite: 13,920 passed, 0 failed.
- Codex: SAFE TO SHIP — change confined to the ordinary transcript branch; all body permutations render card+body exactly once with stable ordering; persisted replay identical; sanitization intact.
- Fable: SHIP-UX (no tweaks). Status-only turns verified byte-identical (no empty body div).

A tool-limit turn's final written report now renders alongside its status card instead of being suppressed by it.
